go/vt/discovery/topology_watcher.go

```diff
@@ -437,3 +437,54 @@ func (fbs *FilterByShard) isIncluded(tablet *topodatapb.Tablet) bool {
 	}
 	return false
 }
+
+// FilterByKeyspace is a TabletRecorder filter that filters tablets by
+// keyspace
+type FilterByKeyspace struct {
+	tr TabletRecorder
+
+	keyspaces map[string]bool
+}
+
+// NewFilterByKeyspace creates a new FilterByKeyspace on top of an existing
+// TabletRecorder. Each filter is a keyspace entry. All tablets that match
+// a keyspace will be forwarded to the underlying TabletRecorder.
+func NewFilterByKeyspace(tr TabletRecorder, selectedKeyspaces []string) *FilterByKeyspace {
+	m := make(map[string]bool)
+	for _, keyspace := range selectedKeyspaces {
+		m[keyspace] = true
+	}
+
+	return &FilterByKeyspace{
+		tr:        tr,
+		keyspaces: m,
+	}
+}
+
+// AddTablet is part of the TabletRecorder interface.
+func (fbk *FilterByKeyspace) AddTablet(tablet *topodatapb.Tablet, name string) {
+	if fbk.isIncluded(tablet) {
+		fbk.tr.AddTablet(tablet, name)
+	}
+}
+
+// RemoveTablet is part of the TabletRecorder interface.
+func (fbk *FilterByKeyspace) RemoveTablet(tablet *topodatapb.Tablet) {
+	if fbk.isIncluded(tablet) {
+		fbk.tr.RemoveTablet(tablet)
+	}
+}
+
+// ReplaceTablet is part of the TabletRecorder interface.
+func (fbk *FilterByKeyspace) ReplaceTablet(old, new *topodatapb.Tablet, name string) {
+	if fbk.isIncluded(old) && fbk.isIncluded(new) {
```

Comment:
The check for: ```fbk.isIncluded(new)``` might be redundant. A tablet shouldn't be able to change keyspaces. So if it was included in the old it should be included in new.

Reply:
In that case, maybe we should check for those failure conditions - `isIncluded(new) && !isIncluded(old)` and the reverse - and return a `FAILED_PRECONDITION` error.

Reply:
> if it was included in the old it should be included in new.

☝️ I'm probably missing some important context here -- can you expand more about what you mean here? My understanding is that given that tablets can't change keyspaces, the `fbk.isIncluded(old)` is the unnecessary check here (as opposed to `fbk.isIncluded(new)`).

Reply:
Looks like I posted the above comment on a non-refreshed page - just now seeing @deepthi's comment.

@deepthi it looks like the callers of `ReplaceTablet` do not have mechanisms for handling these types of errors. Do we know what the expected behavior for FAILED_PRECONDITION errors during tablet replacement is?

Given the low likelihood of us hitting this case, would logging be sufficient here?

Reply:
Oh I see.. the interface won't let you return an error.
Yes, logging it would be fine.